### ISSUE TITLE
Memory leaks with etcdv3 mode confd v0.14.0 (13 too)

### ISSUE THREAD

**Issue by Zhiz0id** [kelseyhightower/confd#615]
Runs like:
`./confd-0.14.0-linux-amd64 -backend etcdv3 -srv-record _etcd-client-ssl._tcp.dev.lan -client-ca-keys /etc/confd/certs/ca.pem -client-cert /etc/confd/certs/etcd-client.pem -client-key=/etc/confd/certs/etcd-client-key.pem -watch`

Memory stat:
```
# while true; do ps -eo pid,vsz,rss,cmd,time -q `pidof confd-0.14.0-linux-amd64`; sleep 60; done
  PID    VSZ   RSS CMD                             TIME
25809  27064 16024 ./confd-0.14.0-linux-amd64  00:00:00
  PID    VSZ   RSS CMD                             TIME
25809  45800 34420 ./confd-0.14.0-linux-amd64  00:00:01
  PID    VSZ   RSS CMD                             TIME
25809  89328 78388 ./confd-0.14.0-linux-amd64  00:00:03
  PID    VSZ   RSS CMD                             TIME
25809 157232 145556 ./confd-0.14.0-linux-amd64  00:00:05
  PID    VSZ   RSS CMD                             TIME
25809 260424 249392 ./confd-0.14.0-linux-amd64  00:00:07
```

```
./confd-0.14.0-linux-amd64 -version
confd 0.14.0 (Git SHA: 9fab9634, Go Version: go1.9.1)

etcd --version
etcd Version: 3.1.7
Git SHA: 43b7507
Go Version: go1.7.4
Go OS/Arch: linux/amd64
```

**Comment by Zhiz0id** [comment]
Looks like the problem somewhere in the github.com/coreos/etcd/clientv3

> Make sure to close the client after using it. If the client is not closed, the connection will have leaky goroutines.

I have patched confd and counted a goroutines and after 100+ restarted processor.Process() it helps a little. 
As I understand there must be a client reconnect or something like that after some time, but I don't know how to do that.

```
while true; do ps -eo pid,vsz,rss,cmd,time -q `pidof confd`; sleep 60; done
  PID    VSZ   RSS CMD                             TIME
29589 521816 19612 ./confd -backend etcdv3 -sr 00:00:00
  PID    VSZ   RSS CMD                             TIME
29589 769176 38684 ./confd -backend etcdv3 -sr 00:00:01 <-- 100 goroutines
  PID    VSZ   RSS CMD                             TIME
29589 809536 78764 ./confd -backend etcdv3 -sr 00:00:02
  PID    VSZ   RSS CMD                             TIME
29589 853288 122952 ./confd -backend etcdv3 -sr 00:00:04
  PID    VSZ   RSS CMD                             TIME
29589 894732 164824 ./confd -backend etcdv3 -sr 00:00:06
```

164mb vs 249mb is better but still bad.

**Comment by zyf0330** [comment]
I have test and find after one hour, vsz and rss don't increase anymore.
```
Tue Oct 31 13:04:51 CST 2017
 2805 308912 13960 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:05:51 CST 2017
 2805 308912 13960 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:06:51 CST 2017
 2805 308912 13596 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:07:51 CST 2017
 2805 308912 13596 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:08:51 CST 2017
 2805 308912 13648 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:09:51 CST 2017
 2805 308912 12356 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:10:51 CST 2017
 2805 460472 14696 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:11:51 CST 2017
 2805 460472 14696 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:12:51 CST 2017
 2805 460472 14504 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:13:51 CST 2017
 2805 460472 16104 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:14:51 CST 2017
 2805 460472 16104 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:15:51 CST 2017
 2805 460472 15928 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:16:51 CST 2017
 2805 460472 15928 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:17:51 CST 2017
 2805 460472 15928 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:18:51 CST 2017
 2805 460472 16104 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:19:51 CST 2017
 2805 460472 16104 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:20:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:21:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:22:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:23:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:24:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:25:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:26:51 CST 2017
 2805 526008 16168 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:27:51 CST 2017
 2805 526008 15552 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:28:51 CST 2017
 2805 526008 16136 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:29:51 CST 2017
 2805 526008 16136 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:30:51 CST 2017
 2805 526008 16172 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:31:51 CST 2017
 2805 526008 16172 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:32:51 CST 2017
 2805 526008 16172 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:33:51 CST 2017
 2805 526008 16172 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:34:51 CST 2017
 2805 526008 16172 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:36:51 CST 2017
 2805 526008 15556 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:37:51 CST 2017
 2805 526008 16148 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:38:51 CST 2017
 2805 526008 16148 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:39:51 CST 2017
 2805 526008 16148 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:40:51 CST 2017
 2805 591544 16188 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:41:52 CST 2017
 2805 591544 16188 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:42:52 CST 2017
 2805 591544 16188 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:43:52 CST 2017
 2805 591544 16188 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:44:52 CST 2017
 2805 591544 16188 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:45:52 CST 2017
 2805 591544 15620 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:46:52 CST 2017
 2805 591544 15620 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:47:52 CST 2017
 2805 591544 16156 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:48:52 CST 2017
 2805 591544 16156 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:49:52 CST 2017
 2805 591544 16156 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:50:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:51:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:52:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:53:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:54:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:55:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:56:52 CST 2017
 2805 591544 16196 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:57:52 CST 2017
 2805 591544 15700 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:58:52 CST 2017
 2805 591544 15700 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 13:59:52 CST 2017
 2805 591544 15700 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:00:52 CST 2017
 2805 591544 16108 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:01:52 CST 2017
 2805 591544 16220 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:02:52 CST 2017
 2805 591544 16220 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:03:52 CST 2017
 2805 591544 16108 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:04:52 CST 2017
 2805 591544 16108 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:05:52 CST 2017
 2805 591544 16220 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:06:52 CST 2017
 2805 591544 15916 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:07:52 CST 2017
 2805 591544 15916 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:08:52 CST 2017
 2805 591544 15916 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:09:52 CST 2017
 2805 591544 15852 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:10:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:11:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:12:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:13:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:14:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:15:52 CST 2017
 2805 591544 15800 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:16:52 CST 2017
 2805 591544 15800 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:17:52 CST 2017
 2805 591544 15800 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:18:52 CST 2017
 2805 591544 15768 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:19:52 CST 2017
 2805 591544 15768 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:20:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:21:52 CST 2017
 2805 591544 15848 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:22:52 CST 2017
 2805 591544 15848 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:23:52 CST 2017
 2805 591544 15848 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:24:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:25:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:26:52 CST 2017
 2805 591544 16232 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:27:52 CST 2017
 2805 591544 15400 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:28:52 CST 2017
 2805 591544 15400 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:29:52 CST 2017
 2805 591544 16200 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:30:52 CST 2017
 2805 591544 16268 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:31:52 CST 2017
 2805 591544 16272 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:32:52 CST 2017
 2805 591544 16272 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:33:52 CST 2017
 2805 591544 16272 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:34:52 CST 2017
 2805 591544 16272 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:35:52 CST 2017
 2805 591544 16272 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:36:52 CST 2017
 2805 591544 16088 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:37:52 CST 2017
 2805 591544 16088 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:38:52 CST 2017
 2805 591544 16088 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:39:52 CST 2017
 2805 591544 15504 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:40:52 CST 2017
 2805 591544 15504 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:41:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:42:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:43:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:44:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:45:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:46:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:47:52 CST 2017
 2805 591544 15956 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:48:52 CST 2017
 2805 591544 15584 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:49:52 CST 2017
 2805 591544 15584 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:50:52 CST 2017
 2805 591544 15584 /usr/local/bin/confd -confi 00:00:00
Tue Oct 31 14:51:52 CST 2017
 2805 591544 15680 /usr/local/bin/confd -confi 00:00:00
```

**Comment by zyf0330** [comment]
@Zhiz0id I use etcdv3 too.
Cmd: 
```
/usr/local/bin/confd -config-file /etc/confd/confd.toml -basic-auth -username *** -password *** -watch
```
toml: 
```
backend = "etcdv3"
confdir = "/etc/confd/kcp"
nodes = [
  "1",
  "2",
  "3",
]
```
I think maybe we have different situation. I don't often change config.

**Comment by zyf0330** [comment]
I find something by go pprof. 
I change code in `WatchPrefix` into `context.WithTimeout` with 100 milliseconds to make this process loop quickly. Goroutines keeps increasing like this:
```
goroutine profile: total 165
37 @ 0x42edac 0x43f1c8 0x558aa3 0x45e291
#	0x558aa2	context.propagateCancel.func1+0x112	/usr/local/go/src/context/context.go:260

37 @ 0x42edac 0x43f1c8 0xa8349c 0xa8418c 0x4759c6 0x475b38 0xa8ee29 0xa8f93d 0xa99524 0xabd332 0xb21fce 0x45e291
#	0xa8349b	github.com/zyf0330/confd/vendor/google.golang.org/grpc/transport.(*recvBufferReader).Read+0x2eb			/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/transport/transport.go:143
#	0xa8418b	github.com/zyf0330/confd/vendor/google.golang.org/grpc/transport.(*Stream).Read+0x5b				/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/transport/transport.go:333
#	0x4759c5	io.ReadAtLeast+0x85												/usr/local/go/src/io/io.go:309
#	0x475b37	io.ReadFull+0x57												/usr/local/go/src/io/io.go:327
#	0xa8ee28	github.com/zyf0330/confd/vendor/google.golang.org/grpc.(*parser).recvMsg+0x68					/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/rpc_util.go:243
#	0xa8f93c	github.com/zyf0330/confd/vendor/google.golang.org/grpc.recv+0x4c						/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/rpc_util.go:339
#	0xa99523	github.com/zyf0330/confd/vendor/google.golang.org/grpc.(*clientStream).RecvMsg+0x113				/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/stream.go:386
#	0xabd331	github.com/zyf0330/confd/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb.(*watchWatchClient).Recv+0x61	/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:2473
#	0xb21fcd	github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveWatchClient+0x4d	/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3/watch.go:557

37 @ 0x42edac 0x43f1c8 0xa9dc11 0x45e291
#	0xa9dc10	github.com/zyf0330/confd/vendor/google.golang.org/grpc.newClientStream.func3+0x1a0	/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/google.golang.org/grpc/stream.go:236

37 @ 0x42edac 0x43f1c8 0xb2098d 0x45e291
#	0xb2098c	github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3.(*watchGrpcStream).run+0x3ec	/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3/watch.go:430
```
It shows that `WatchPrefix` called repeatly causes more and more repeat goroutines, and heap usage becomes more and more with it.

**Comment by okushchenko** [comment]
It looks like the `Watcher` from Etcd v3 client must be closed properly at the end of `WatchPrefix` method, in contrast to Etcd v2 client, which has no close method for `Watched`. https://godoc.org/github.com/coreos/etcd/clientv3#Watcher

**Comment by zyf0330** [comment]
I attempted to add `Watcher.Close()` in the end of `WatchPrefix`, but I found heap usage keeps increasing.
I don't know too much about go pprof, so could you see it? Thanks.

**Comment by zyf0330** [comment]
I think we cannot call `Watcher.Close()` simplely in `WatchPrefix`, keepalive process doesn't work anymore after calling it. Maybe you should change code logic of `watch` and I cannot help a lot.

**Comment by zyf0330** [comment]
@okushchenko Sorry, I make a mistake.
When I use `WIthTimeout` context, etcd starts a new `newWatcherGrpcStream` every time. But actually, in confd code, it should test with `time.Sleep(100 * time.Millisecond); return;`, and this time I get this situation of goroutines.
```
43 @ 0x432ddd 0x443890 0xdd6d1e 0x463ab1
#	0xdd6d1d	github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream+0x39d	/home/zyf/project/go/src/github.com/zyf0330/confd/vendor/github.com/coreos/etcd/clientv3/watch.go:603
```
You can see, this time just one type of goroutine keeping increasing. Memory leak still exists.

I don't want to change code logic, so I change to reuse previous watch channel to avoid memory leak.
I have issue a #631.

**Comment by okushchenko** [comment]
Well done, I've made a code review in #631.

**Comment by zyf0330** [comment]
@okushchenko I just find a bug of #631.
I use `prefix` as key to distinguish different watch. If two config in `conf.d` have the same `prefix` but any different one of `src`, `dest` or `keys`, mistake will appear.
To solve it, key needs to be `src+dest+prefix` to ensure it is the only one. And definition of `WatchPrefix` of `StoreClient` need to be changed. How do you think?

**Comment by okushchenko** [comment]
@zyf0330 I've opened a PR #636 with a fix, check it out. I've updated etcd to v3.2.10, so it might fix #628 too.
